util/oci/client.go

```diff
@@ -0,0 +1,302 @@
+	TestRepo(ctx context.Context) (bool, error)
+}
+
+type Creds struct {
+	Username           string
+	Password           string
+	CAPath             string
+	CertData           []byte
+	KeyData            []byte
+	InsecureSkipVerify bool
+	InsecureHttpOnly   bool
+}
+
+type ClientOpts func(c *nativeOciClient)
+
+func WithIndexCache(indexCache indexCache) ClientOpts {
+	return func(c *nativeOciClient) {
+		c.indexCache = indexCache
+	}
+}
+
+func WithChartPaths(repoCachePaths argoio.TempPaths) ClientOpts {
+	return func(c *nativeOciClient) {
+		c.repoCachePaths = repoCachePaths
+	}
+}
+func NewClient(repoURL string, creds Creds, proxy string, opts ...ClientOpts) (Client, error) {
+	return NewClientWithLock(repoURL, creds, globalLock, proxy, opts...)
+}
+func NewClientWithLock(repoURL string, creds Creds, repoLock sync.KeyLock, proxyUrl string, opts ...ClientOpts) (Client, error) {
+	ociRepo := strings.TrimPrefix(repoURL, "oci://")
+	repo, err := remote.NewRepository(ociRepo)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize repository: %v", err)
+	}
+
+	repo.PlainHTTP = creds.InsecureHttpOnly
+
+	ociUri, err := url.Parse(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse registry host: %v", err)
+	}
+
+	var tlsConf *tls.Config
+	if !repo.PlainHTTP {
+		tlsConf, err = newTLSConfig(creds)
+		if err != nil {
+			return nil, fmt.Errorf("failed setup tlsConfig: %v", err)
+		}
+	}
+
+	var ociRegistry string
+	if ociUri.Port() != "" {
+		ociRegistry = fmt.Sprintf("%s:%s", ociUri.Hostname(), ociUri.Port())
+	} else {
+		ociRegistry = ociUri.Hostname()
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		Proxy:             proxy.GetCallback(proxyUrl),
+		TLSClientConfig:   tlsConf,
+		DisableKeepAlives: true,
+	}}
+	repo.Client = &auth.Client{
+		Client: client,
+		Cache:  nil,
+		Credential: auth.StaticCredential(ociRegistry, auth.Credential{
+			Username: creds.Username,
+			Password: creds.Password,
+		}),
+	}
+
+	c := &nativeOciClient{
+		creds:    creds,
+		repoURL:  ociRepo,
+		proxy:    proxyUrl,
+		repoLock: repoLock,
+		repo:     repo,
+	}
+	for i := range opts {
+		opts[i](c)
+	}
+	return c, nil
+}
+
+// nativeOciClient implements Client interface using oras-go
+type nativeOciClient struct {
+	creds          Creds
+	repoURL        string
+	proxy          string
+	repo           *remote.Repository
+	repoLock       sync.KeyLock
+	indexCache     indexCache
+	repoCachePaths argoio.TempPaths
+}
+
+func (c *nativeOciClient) TestRepo(ctx context.Context) (bool, error) {
+	err := c.repo.Tags(ctx, "", func(tags []string) error {
+		return nil
+	})
+	return err == nil, err
+}
+
+func (c *nativeOciClient) Extract(ctx context.Context, revision string, project string, manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool) (string, argoio.Closer, error) {
```

Comment:
@blakepettersson , do we need `manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool`?

Reply:
Currently no - but for Helm there was a CVE (GHSA-g687-f2gx-6wm8), which didn't have any upper bound on the size of the extracted files. Ideally there should be something similar for OCI as well; I presume there is a built in way in ORAS for us to limit the max size of the downloaded manifests, but I haven't looked at what that could be yet.

Reply:
Can someone help me to understand the expected behavior of "Extracting"? By performing extracting, do we want to download just artifact files, or both the manifest and the files?

> I presume there is a built in way in ORAS for us to limit the max size of the downloaded manifests
Do we want the limit the max size of the manifest downloaded to disk, or in memory?

Reply:
> By performing extracting, do we want to download just artifact files, or both the manifest and the files?

At this time it would only be the files.

>  Do we want the limit the max size of the manifest downloaded to disk, or in memory?

I was thinking of to disk, but if it is required also to memory.

Reply:
If the manifest wouldn't be downloaded, we just need to limit its size in memory. The default copy options have a 4-MB limit on manifest size. See [`CopyGraphOptions.MaxMetadataBytes`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0#CopyGraphOptions).